onnxruntime/core/framework/kernel_registry.cc

```diff
@@ -1,20 +1,23 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include <memory>
 #include <unordered_map>
 
 #include "core/framework/kernel_registry.h"
 
 using namespace ::onnxruntime::common;
 namespace onnxruntime {
 
-// map from input/output name or type string to TypeProto pointer
-using TypeBindingMap = std::unordered_map<std::string, const ONNX_NAMESPACE::TypeProto*>;
+namespace {
+// Function that is called for each formal parameter and its associated TypeProto.
+// Returns true if traversal should continue, false otherwise.
+using TraverseFn = std::function<bool(const ONNX_NAMESPACE::OpSchema::FormalParameter&,
+                                      const ONNX_NAMESPACE::TypeProto*)>;
 
-// Builds a map from name to bound type for the specified node.
-// The name can represent either a type parameter or an input/output parameter.
-TypeBindingMap CreateTypeBindingMap(const onnxruntime::Node& node) {
-  TypeBindingMap result{};
+// Traverses the node's formal parameters and calls the provided function with the formal parameter and its associated
+// TypeProto.
+void TraverseFormalParametersWithTypeProto(const onnxruntime::Node& node, const TraverseFn& traverse_fn) {
```

Comment:
A simpler solution would be:

```c++
template <typename T>
void TraverseFormalParametersWithTypeProto(const onnxruntime::Node& node, const T& traverse_fn){
   ...
}
```

It's faster because it can avoid encapsulating a function pointer(or lambda) to a std::function object.

Reply:
I think std::function is better for async operations, which stores the information for future use.

Reply:
Good point, will update.